### PR TITLE
Add the Elasticsearch vectorstore

### PR DESCRIPTION
**Pull request by ezimuel**
This PR adds the support of Elasticsearch as vectorstore. It includes unit tests, documentation and examples.

### CODE REVIEW

File: examples/src/indexes/vector_stores/elasticsearch/elasticsearch.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import { Client } from "@elastic/elasticsearch";
+import { OpenAIEmbeddings } from "langchain/embeddings/openai";
+import { ElasticsearchVectorStore } from "langchain/vectorstores/elasticsearch";
+
+// to run this first run chroma's docker-container with `docker-compose up -d --build`
+export async function run() {
+  const client = new Client({
+    nodes: [process.env.ELASTICSEARCH_URL ?? "http://127.0.0.1:9200"],
+    auth: {
+      username: "elastic",
```
Comment (cdiddy77) [inline examples/src/indexes/vector_stores/elasticsearch/elasticsearch.ts:10]: maybe have this be an environment variable as well?

File: langchain/.env.example [hunk 2]
```diff
@@ -9,6 +9,7 @@ AZURE_OPENAI_API_DEPLOYMENT_NAME=ADD_YOURS_HERE
 AZURE_OPENAI_API_COMPLETIONS_DEPLOYMENT_NAME=ADD_YOURS_HERE
 AZURE_OPENAI_API_EMBEDDINGS_DEPLOYMENT_NAME=ADD_YOURS_HERE
 AZURE_OPENAI_API_VERSION=ADD_YOURS_HERE
+ELASTICSEARCH_URL=http://127.0.0.1:9200
```
Comment (cdiddy77) [inline langchain/.env.example:12]: also ELASTICSEARCH_PASSWORD and maybe ELASTICSEARCH_USERNAME

File: langchain/src/vectorstores/elasticsearch.ts [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    efConstruction = 100,
+    m = 16
+  ): Promise<void> {
+    const request: estypes.IndicesCreateRequest = {
+      index: this.indexName,
+      mappings: {
+        dynamic_templates: [
+          {
+            // map all metadata properties to be keyword
+            "metadata.*": {
```
Comment (cdiddy77) [inline langchain/src/vectorstores/elasticsearch.ts:162]: ah clever. My solution was more complicated.

File: langchain/src/vectorstores/tests/elasticsearch.int.test.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
+/* eslint-disable no-process-env */
```
Comment (cdiddy77) [inline langchain/src/vectorstores/tests/elasticsearch.int.test.ts:1]: consider: also incorporating [unit tests](https://github.com/hwchase17/langchainjs/pull/1718/files#diff-e4c8fc644431432bfd8f99ebce64d13484f8434c0180769af36cc1d023654fd7R1) as well as [tests that involve metadata lookup](https://github.com/hwchase17/langchainjs/pull/1718/files#diff-4b771d4d6505aae476d14c91b8496fbf5ae591858597f482e30a77cc70a0bb8eR36)

File: langchain/src/vectorstores/elasticsearch.ts [hunk 5]
```diff
@@ -0,0 +1,213 @@
+import { Client } from "@elastic/elasticsearch";
+import { estypes } from "@elastic/elasticsearch";
```
Comment (JoshMock) [inline langchain/src/vectorstores/elasticsearch.ts:2]:
```suggestion
import { Client, estypes } from "@elastic/elasticsearch";
```

**Review by JoshMock — COMMENTED** [review]
From an Elasticsearch client usage perspective this looks great. :+1:

File: docs/docs/modules/indexes/vector_stores/integrations/elasticsearch.md [hunk 6]
```diff
@@ -0,0 +1,132 @@
+---
```
[the remark this answers is not in the document]
Comment (jacoblee93) [reply docs/docs/modules/indexes/vector_stores/integrations/elasticsearch.md:1]: It looks like it requires a Node-only dependency, at least by the npm page: https://www.npmjs.com/package/@elastic/elasticsearch

**Comment by jacoblee93** [comment]
Thank you, and apologies for the delay! Will review this one instead of #1718

File: docs/docs/modules/indexes/vector_stores/integrations/elasticsearch.md [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+import { Client } from "@elastic/elasticsearch";
+import { VectorDBQAChain } from "langchain/chains";
+import { OpenAIEmbeddings } from "langchain/embeddings/openai";
+import { OpenAI } from "langchain/llms/openai";
+import { ElasticsearchVectorStore } from "langchain/vectorstores/elasticsearch";
+
+const config: any = {
+  node: process.env.ELASTICSEARCH_URL,
+};
+if (process.env.ELASTICSEARCH_API_KEY) {
```
Comment (jacoblee93) [inline docs/docs/modules/indexes/vector_stores/integrations/elasticsearch.md:84]: You can use an API key in lieu of a username/password combination then?

File: langchain/src/vectorstores/elasticsearch.ts [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+
+export interface ElasticsearchClientArgs {
+  readonly client: Client;
+  readonly indexName?: string;
+  readonly vectorSearchOptions?: VectorSearchOptions;
+}
+
+type ElasticsearchFilter = object;
+
+export class ElasticsearchVectorStore extends VectorStore {
```
Comment (jacoblee93) [inline langchain/src/vectorstores/elasticsearch.ts:25]: In Python this seems to be called ElasticVectorSearch: https://python.langchain.com/docs/modules/data_connection/vectorstores/integrations/elasticsearch

Not sure if you prefer the synchronization or if this is a better name now?

File: docs/docs/modules/indexes/vector_stores/integrations/elasticsearch.md [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+```typescript
+import { Client } from "@elastic/elasticsearch";
+import { Document } from "langchain/document";
+import { OpenAIEmbeddings } from "langchain/embeddings/openai";
+import { ElasticsearchVectorStore } from "langchain/vectorstores/elasticsearch";
+
+const config: any = {
+  node: process.env.ELASTICSEARCH_URL,
+};
+if (process.env.ELASTICSEARCH_API_KEY) {
```
Comment (jacoblee93) [inline docs/docs/modules/indexes/vector_stores/integrations/elasticsearch.md:36]: How do I create an API key?

File: docs/docs/modules/indexes/vector_stores/integrations/elasticsearch.md (hunk 6 above)
Comment (jacoblee93) [inline docs/docs/modules/indexes/vector_stores/integrations/elasticsearch.md:1]: Best to make this a `.mdx` file and import code snippets from `examples/` but I'm happy to do this later

File: docs/docs/modules/indexes/vector_stores/integrations/elasticsearch.md (hunk 9 above)
Comment (jacoblee93) [reply docs/docs/modules/indexes/vector_stores/integrations/elasticsearch.md:36]: (Trying out the cloud version)

**Comment by jacoblee93** [comment]
Broadly LGTM, seems like you have a build failure though. I don't think I have permissions to push directly to your branch but I can open an extension PR tomorrow to fix the build and change docs to an MDX file.

File: examples/src/indexes/vector_stores/elasticsearch/elasticsearch.ts [hunk 10]
```diff
@@ -0,0 +1,28 @@
+import { Client } from "@elastic/elasticsearch";
```
Comment (jacoblee93) [inline examples/src/indexes/vector_stores/elasticsearch/elasticsearch.ts:1]: We'll also need to add this to `examples/package.json`

**Comment by jacoblee93** [comment]
There's a linter warning, will open another PR extending this and will aim to get it merged for tomorrow's release!

**Comment by ezimuel** [comment]
The linter warning is reported in Vercel? Unfortunately, I cannot access it (404). In the github actions I cannot find it.

**Comment by jacoblee93** [comment]
All good, I opened #1810 with some small fixes and will merge that. Thanks again!

**Comment by jacoblee93** [comment]
Merged in #1810!

**Pull request closed without merging**